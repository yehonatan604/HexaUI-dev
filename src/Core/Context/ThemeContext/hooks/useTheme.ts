import { useContext } from "react";
import { ThemeContext } from "../Theme.context";

const useTheme = () => {
    const context = useContext(ThemeContext);
    if (context === undefined) {
        throw new Error("useTheme must be used within a ThemeProvider");
    }
    return {
        mode: context.mode,
        toggleTheme: context.toggleTheme,
        colors: context.colors,
        setColors: context.setColors,
        gradients: context.gradients,
        setGradients: context.setGradients
    };
};

export default useTheme;
