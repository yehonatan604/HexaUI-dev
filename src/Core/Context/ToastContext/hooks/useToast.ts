import { useContext } from "react";
import { ToastStackContext } from "../Toast.context";

const useToast = () => {
    const context = useContext(ToastStackContext);
    if (context === undefined) {
        throw new Error("useTheme must be used within a ThemeProvider");
    }
    return {
        addToast: context.addToast,
        removeToast: context.removeToast,
        toast: context.toast
    };
};
export default useToast;