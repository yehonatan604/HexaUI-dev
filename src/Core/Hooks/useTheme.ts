import { useContext } from "react";
import { ThemeContext } from "../Context/Theme.context";

const useTheme = () => {
    const context = useContext(ThemeContext);
    if (context === undefined) {
        throw new Error("useTheme must be used within a ThemeProvider");
    }
    return context;
};
export default useTheme;