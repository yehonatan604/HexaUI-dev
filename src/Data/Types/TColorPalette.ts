import { TGradientPalette } from "./TGradientPalette";

export type TColorPalette = TGradientPalette & {
    black?: string;
    white?: string;
    backgroundLight?: string;
    backgroundDark?: string;
    textLight?: string;
    textDark?: string;

    primaryLight?: string;
    secondaryLight?: string;
    standradLight?: string;
    infoLight?: string;
    selectLight?: string;
    cancelLight?: string;
    approveLight?: string;
    successLight?: string;
    warningLight?: string;
    failureLight?: string;
    blackLight?: string;
    whiteLight?: string;

    primaryDark?: string;
    secondaryDark?: string;
    standradDark?: string;
    infoDark?: string;
    selectDark?: string;
    cancelDark?: string;
    approveDark?: string;
    successDark?: string;
    warningDark?: string;
    failureDark?: string;
    blackDark?: string;
    whiteDark?: string;
}
