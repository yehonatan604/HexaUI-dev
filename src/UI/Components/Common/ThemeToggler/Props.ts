import { TComponent } from "../../../../Data/Types/ComponentTypes/TComponent";

export type ThemeTogglerProps = TComponent<"button"> & {
    darkIcon?: React.ReactNode;
    lightIcon?: React.ReactNode;
};