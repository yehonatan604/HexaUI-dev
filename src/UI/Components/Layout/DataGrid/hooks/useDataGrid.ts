import React, { useState, useMemo } from "react";
import { TFilterConfig } from "../types/TFilterConfig";
import { TSortConfig } from "../types/TSortConfig";

const useDataGrid = (rowsArr: React.ReactNode[]) => {
    const [searchTerm, setSearchTerm] = useState("");
    const [sortConfig, setSortConfig] = useState<TSortConfig>({
        key: -1,
        direction: null,
    });
    const [filterConfig, setFilterConfig] = useState<TFilterConfig>({
        column: 0,
        operator: "equals",
        value: "",
    });

    const dataRows = rowsArr.slice(1);

    // Add original indices to data rows
    const indexedDataRows = useMemo(() => {
        return dataRows.map((row) => ({
            data: row,
        }));
    }, [dataRows]);

    // Filter and sort the data rows
    const filteredAndSortedRows = useMemo(() => {
        const applyFilter = (row: { data: React.ReactNode[] }) => {
            const cellValue = React.Children.toArray(row.data)[filterConfig.column];
            const cellText = React.isValidElement(cellValue)
                ? (cellValue.props.children?.toString() || "").toLowerCase()
                : String(cellValue).toLowerCase();
            const filterValue = filterConfig.value.toLowerCase();

            switch (filterConfig.operator) {
                case "equals":
                    return cellText === filterValue;
                case "contains":
                    return cellText.includes(filterValue);
                case "greater":
                    return Number(cellText) > Number(filterValue);
                case "less":
                    return Number(cellText) < Number(filterValue);
                case "regex":
                    try {
                        const regex = new RegExp(filterConfig.value, "i");
                        return regex.test(cellText);
                    } catch {
                        return true;
                    }
                default:
                    return true;
            }
        };
        // First apply search filter
        let filteredRows = indexedDataRows;
        if (searchTerm) {
            filteredRows = filteredRows.filter((row) => {
                return React.Children.toArray(row.data).some((cell) => {
                    const cellText = React.isValidElement(cell)
                        ? (cell.props.children?.toString() || "").toLowerCase()
                        : String(cell).toLowerCase();
                    return cellText.includes(searchTerm.toLowerCase());
                });
            });
        }

        // Then apply column filter
        if (filterConfig.value) {
            filteredRows = filteredRows.filter(applyFilter);
        }

        // Then sort
        if (sortConfig.direction === null || sortConfig.key === -1) {
            return filteredRows;
        }

        return [...filteredRows].sort((a, b) => {
            const aValue = React.Children.toArray(a.data)[sortConfig.key];
            const bValue = React.Children.toArray(b.data)[sortConfig.key];

            const aText = React.isValidElement(aValue)
                ? aValue.props.children?.toString() || ""
                : String(aValue);
            const bText = React.isValidElement(bValue)
                ? bValue.props.children?.toString() || ""
                : String(bValue);

            const aNum = Number(aText);
            const bNum = Number(bText);

            if (!isNaN(aNum) && !isNaN(bNum)) {
                return sortConfig.direction === "asc" ? aNum - bNum : bNum - aNum;
            }

            return sortConfig.direction === "asc"
                ? aText.localeCompare(bText)
                : bText.localeCompare(aText);
        });
    }, [indexedDataRows, searchTerm, filterConfig, sortConfig]);

    return {
        filteredAndSortedRows,
        sortConfig,
        setSortConfig,
        searchTerm,
        setSearchTerm,
        filterConfig,
        setFilterConfig,
    };
};

export default useDataGrid;