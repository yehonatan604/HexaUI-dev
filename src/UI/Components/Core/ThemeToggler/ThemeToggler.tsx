import { useEffect, useState } from "react";
import useTheme from "../../../../Core/Context/ThemeContext/hooks/useTheme";
import { TThemeToggler } from "./types/TThemeToggler";

const ThemeToggler = (props: TThemeToggler) => {
  const { darkIcon, lightIcon, className, ...componentProps } = props;
  const [icon, setIcon] = useState<React.ReactNode>(darkIcon);
  const { mode, toggleTheme } = useTheme();

  useEffect(() => {
    if (lightIcon && darkIcon) setIcon(mode === "light" ? darkIcon : lightIcon);
    else {
      if (darkIcon) setIcon(mode === "light" ? darkIcon : "☀️");
      else if (lightIcon) setIcon(mode === "light" ? "🌙" : lightIcon);
      else setIcon(mode === "light" ? "🌙" : "☀️");
    }
  }, [mode, darkIcon, lightIcon]);

  return (
    <button onClick={toggleTheme} {...componentProps} className={`${className}`}>
      {icon}
    </button>
  );
};

export default ThemeToggler;
