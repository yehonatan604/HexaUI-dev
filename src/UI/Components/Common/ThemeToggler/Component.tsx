import { useEffect, useState } from "react";
import useTheme from "../../../../Core/Hooks/useTheme";
import { ThemeTogglerProps } from "./Props";

const ThemeToggler = (props: ThemeTogglerProps) => {
  const { darkIcon, lightIcon, ...componentProps } = props;
  const [icon, setIcon] = useState<React.ReactNode>(darkIcon);
  const { mode, toggleTheme } = useTheme();

  useEffect(() => {
    if (lightIcon && darkIcon) setIcon(mode === "light" ? darkIcon : lightIcon);
    else {
      if (darkIcon) setIcon(mode === "light" ? darkIcon : "☀️");
      else if (lightIcon) setIcon(mode === "light" ? "🌙" : lightIcon);
      else setIcon(mode === "light" ? "🌙" : "☀️");
    }
  }, [mode, darkIcon, lightIcon]);

  return (
    <button onClick={toggleTheme} {...componentProps}>
      {icon}
    </button>
  );
};

export default ThemeToggler;
