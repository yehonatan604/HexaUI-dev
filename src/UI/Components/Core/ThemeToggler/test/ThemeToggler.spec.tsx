import { render, screen, fireEvent } from "@testing-library/react";
import ThemeToggler from "../ThemeToggler";
import { describe, beforeEach, it, vi, expect } from "vitest";
import * as themeContext from "../../../../../Core/Context/ThemeContext/hooks/useTheme"; // Import the whole module
import { TColorPalette } from "../../../../../Data/Types/TColorPalette";

// Mock the `useTheme` hook
vi.mock("../../../../../Core/Context/ThemeContext/hooks/useTheme"); // Correctly mock the module

describe("ThemeToggler Component", () => {
  const mockToggleTheme = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();

    // Mock the return value of the `useTheme` hook within the mocked module
    vi.spyOn(themeContext, "default").mockReturnValue({
      mode: "light",
      toggleTheme: mockToggleTheme,
      colors: {} as TColorPalette, // Mock other properties as empty objects or relevant mock data
      setColors: vi.fn(),
      gradients: {} as TColorPalette,
      setGradients: vi.fn(),
    });
  });

  it("renders the correct icon for light mode", () => {
    render(<ThemeToggler darkIcon={<span>🌑</span>} lightIcon={<span>🌞</span>} />);
    expect(screen.getByText("🌑")).toBeInTheDocument();
  });

  it("renders the correct icon for dark mode", () => {
    vi.spyOn(themeContext, "default").mockReturnValue({
      mode: "dark",
      toggleTheme: mockToggleTheme,
      colors: {} as TColorPalette,
      setColors: vi.fn(),
      gradients: {} as TColorPalette,
      setGradients: vi.fn(),
    });

    render(<ThemeToggler darkIcon={<span>🌑</span>} lightIcon={<span>🌞</span>} />);
    expect(screen.getByText("🌞")).toBeInTheDocument();
  });

  it("toggles the theme mode when clicked", () => {
    render(<ThemeToggler darkIcon={<span>🌑</span>} lightIcon={<span>🌞</span>} />);
    const button = screen.getByRole("button");
    fireEvent.click(button);
    expect(mockToggleTheme).toHaveBeenCalledTimes(1);
  });

  it("renders default icons if none are provided", () => {
    render(<ThemeToggler />);
    expect(screen.getByText("🌙")).toBeInTheDocument();

    vi.spyOn(themeContext, "default").mockReturnValue({
      mode: "dark",
      toggleTheme: mockToggleTheme,
      colors: {} as TColorPalette,
      setColors: vi.fn(),
      gradients: {} as TColorPalette,
      setGradients: vi.fn(),
    });

    render(<ThemeToggler />);
    expect(screen.getByText("☀️")).toBeInTheDocument();
  });

  // Snapshot tests to verify output consistency

  it("matches the snapshot for light mode", () => {
    const { asFragment } = render(
      <ThemeToggler darkIcon={<span>🌑</span>} lightIcon={<span>🌞</span>} />
    );
    expect(asFragment()).toMatchSnapshot(); // Snapshot for light mode
  });

  it("matches the snapshot for dark mode", () => {
    vi.spyOn(themeContext, "default").mockReturnValue({
      mode: "dark",
      toggleTheme: mockToggleTheme,
      colors: {} as TColorPalette,
      setColors: vi.fn(),
      gradients: {} as TColorPalette,
      setGradients: vi.fn(),
    });
    const { asFragment } = render(
      <ThemeToggler darkIcon={<span>🌑</span>} lightIcon={<span>🌞</span>} />
    );
    expect(asFragment()).toMatchSnapshot(); // Snapshot for dark mode
  });
});
