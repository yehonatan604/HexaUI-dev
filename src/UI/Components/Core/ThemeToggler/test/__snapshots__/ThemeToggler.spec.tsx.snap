// Vitest Snapshot v1

exports[`ThemeToggler Component > matches the snapshot for dark mode 1`] = `
<DocumentFragment>
  <button>
    <span>
      ðŸŒž
    </span>
  </button>
</DocumentFragment>
`;

exports[`ThemeToggler Component > matches the snapshot for light mode 1`] = `
<DocumentFragment>
  <button>
    <span>
      ðŸŒ‘
    </span>
  </button>
</DocumentFragment>
`;
